### PR TITLE
WebCore::IOSurface should not store a reference to a CGContext

### PR DESCRIPTION
**Pull request by kkinnunen-apple**
#### 902a49a99629ceb2bf363ee4c96b0f3fdd34afc3
<pre>
WebCore::IOSurface should not store a reference to a CGContext
<a href="https://bugs.webkit.org/show_bug.cgi?id=253698">https://bugs.webkit.org/show_bug.cgi?id=253698</a>
rdar://106547690

Reviewed by Matt Woodrow.

An IOSurface might be written to and read from by multiple CGContexts
or even other contexts such as Metal (via WebGL / ANGLE).

An arbitrary context that has been used is not useful anymore in the
future, as the caller will not know the context state.

Due to these points, does not make sense to store a CGContext as a member in
WebCore::IOSurface.

The only real CG draw to an IOSurface happens via ImageBufferIOSurfaceBackend.
Store the context there, as that is the place to use the context.

Caching the context behind the interface creates source of bugs, as in
here fixed WebAutomationSession / ViewSnapshot where the callers
would inadvertedly create the contexts and leave them stored.

Remove invalid comment about CGIOSurfaceContextCreate affecting the
use count. The added test proves otherwise.

* Source/WebCore/platform/graphics/cg/IOSurfacePool.cpp:
(WebCore::IOSurfacePool::willAddSurface):
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp:
(WebCore::ImageBufferIOSurfaceBackend::create):
(WebCore::ImageBufferIOSurfaceBackend::ImageBufferIOSurfaceBackend):
(WebCore::ImageBufferIOSurfaceBackend::~ImageBufferIOSurfaceBackend):
(WebCore::ImageBufferIOSurfaceBackend::context):
(WebCore::ImageBufferIOSurfaceBackend::ensurePlatformContext):
(WebCore::ImageBufferIOSurfaceBackend::copyNativeImage):
(WebCore::ImageBufferIOSurfaceBackend::copyNativeImageForDrawing):
(WebCore::ImageBufferIOSurfaceBackend::sinkIntoNativeImage):
(WebCore::ImageBufferIOSurfaceBackend::releaseGraphicsContext):
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.h:
* Source/WebCore/platform/graphics/cocoa/IOSurface.h:
* Source/WebCore/platform/graphics/cocoa/IOSurface.mm:
(WebCore::IOSurface::create):
(WebCore::IOSurface::createFromImage):
(WebCore::IOSurface::createImage):
(WebCore::IOSurface::sinkIntoImage):
(WebCore::IOSurface::createPlatformContext):
(WebCore::IOSurface::setOwnershipIdentity):
(WebCore::IOSurface::ensurePlatformContext): Deleted.
(WebCore::IOSurface::releasePlatformContext): Deleted.
* Source/WebKit/UIProcess/Automation/cocoa/WebAutomationSessionCocoa.mm:
(WebKit::WebAutomationSession::platformGetBase64EncodedPNGData):
* Source/WebKit/UIProcess/mac/ViewSnapshotStoreMac.mm:
(WebKit::ViewSnapshot::setVolatile):
(WebKit::ViewSnapshot::asImageForTesting):
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferShareableMappedIOSurfaceBackend.cpp:
(WebKit::ImageBufferShareableMappedIOSurfaceBackend::create):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/261699@main">https://commits.webkit.org/261699@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/fed313871a94b69f805bb8ad074b0c73d040d851

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112569 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/21720 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/1227 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/4345 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/121119 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/116635 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/23059 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/12882 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/5494 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/118336 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/17129 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100339 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/105657 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/99075 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/883 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/46168 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/14056 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/919 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/95265 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/14739 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/10293 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/20079 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52924 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8158 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16575 "Built successfully") | | | 
| | | | |

### CODE REVIEW

File: Source/WebCore/platform/graphics/cocoa/IOSurface.mm [hunk 1]
```diff
@@ -337,14 +334,14 @@ static IntSize computeMaximumSurfaceSize()
...
 {
-    return adoptCF(CGIOSurfaceContextCreateImage(ensurePlatformContext()));
+    return adoptCF(CGIOSurfaceContextCreateImage(context));
 }
 
-RetainPtr<CGImageRef> IOSurface::sinkIntoImage(std::unique_ptr<IOSurface> surface)
+RetainPtr<CGImageRef> IOSurface::sinkIntoImage(std::unique_ptr<IOSurface>, RetainPtr<CGContextRef> context)
 {
-    return adoptCF(CGIOSurfaceContextCreateImageReference(surface->ensurePlatformContext()));
+    return adoptCF(CGIOSurfaceContextCreateImageReference(context.get()));
```
Comment (mattwoodrow) [inline Source/WebCore/platform/graphics/cocoa/IOSurface.mm:358]: Can we use `IOSurfaceContextGetSurface` to verify that `context` does indeed wrap this IOSurface?

File: Source/WebCore/platform/graphics/cocoa/IOSurface.h [hunk 2]
```diff
@@ -122,17 +122,17 @@ class IOSurface final {
 
     // Any images created from a surface need to be released before releasing
     // the surface, or an expensive GPU readback can result.
```
Comment (mattwoodrow) [inline Source/WebCore/platform/graphics/cocoa/IOSurface.h]: It's releasing the `CGContext` that will trigger read back of any `CGImages` created using it, right? It was previously this surface, since it was (hopefully, maybe) the only one holding a ref to the `CGContext`.

File: Source/WebCore/platform/graphics/cocoa/IOSurface.h [hunk 3]
```diff
@@ -122,17 +122,17 @@ class IOSurface final {
...
 #ifdef __OBJC__
     id asLayerContents() const { return (__bridge id)m_surface.get(); }
 #endif
     IOSurfaceRef surface() const { return m_surface.get(); }
-    WEBCORE_EXPORT CGContextRef ensurePlatformContext(PlatformDisplayID = 0);
-    // The graphics context cached on the surface counts as a "user", so to get
-    // an accurate result from isInUse(), it needs to be released.
-    WEBCORE_EXPORT void releasePlatformContext();
+
+    WEBCORE_EXPORT RetainPtr<CGContextRef> createPlatformContext(PlatformDisplayID = 0);
```
Comment (mattwoodrow) [inline Source/WebCore/platform/graphics/cocoa/IOSurface.h:137]: What happens if multiple consumers create contexts for the same IOSurface and keep them alive (and both issue draw calls)?

Any created images are only going to guarantee to contain data from their respective context, and any drawing from another context may or may not (racily) be there.

It might be nice to only allow one platform context to be alive at any time, or at least add a comment warning about this?

**Review by mattwoodrow — APPROVED** [review]

File: Source/WebCore/platform/graphics/cocoa/IOSurface.mm (hunk 1 above)
Comment (kkinnunen-apple) [reply Source/WebCore/platform/graphics/cocoa/IOSurface.mm:358]: done

File: Source/WebCore/platform/graphics/cocoa/IOSurface.h (hunk 3 above)
Comment (kkinnunen-apple) [reply Source/WebCore/platform/graphics/cocoa/IOSurface.h:137]: > What happens if multiple consumers create contexts for the same IOSurface and keep them alive (and both issue draw calls)?

The caller is expected to coordinate the mutation of the shared IOSurface. This is what we do when we read and write it both in GPUP and WP.

> It might be nice to only allow one platform context to be alive at any time, or at least add a comment warning about this?

I don't think it's possible / desirable in general. E.g. it's the same if you have a malloc buffer and would say it'd be nice if there would not be multiple CGBitmapContexts pointing to the same malloc buffer.

In other words: the code cannot be such that we would have random places creating random contexts to random IOSurfaces. Using IOSurfaces  needs to be intentional: who drew and with which and flushed when, why would the content be there, etc.
